tests/benchmarks/test_model_schema_generation.py

```diff
@@ -219,3 +238,104 @@ def serialize_model(self) -> Any:
+def test_tagged_union_with_str_discriminator_schema_generation(benchmark):
+    class Cat(BaseModel):
+        pet_type: Literal['cat']
+        meows: int
+
+    class Dog(BaseModel):
+        pet_type: Literal['dog']
+        barks: float
+
+    class Lizard(BaseModel):
+        pet_type: Literal['reptile', 'lizard']
+        scales: bool
+
+    class Model(BaseModel):
+        pet: Union[Cat, Dog, Lizard] = Field(..., discriminator='pet_type')
+        n: int
+
+    benchmark(rebuild_model, Model)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_tagged_union_with_callable_discriminator_schema_generation(benchmark):
+    class Pie(BaseModel):
+        time_to_cook: int
+        num_ingredients: int
+
+    class ApplePie(Pie):
+        fruit: Literal['apple'] = 'apple'
+
+    class PumpkinPie(Pie):
+        filling: Literal['pumpkin'] = 'pumpkin'
+
+    def get_discriminator_value(v: Any) -> str:
+        if isinstance(v, dict):
+            return v.get('fruit', v.get('filling'))
+        return getattr(v, 'fruit', getattr(v, 'filling', None))
+
+    class ThanksgivingDinner(BaseModel):
+        dessert: Annotated[
+            Union[
+                Annotated[ApplePie, Tag('apple')],
+                Annotated[PumpkinPie, Tag('pumpkin')],
+            ],
+            Discriminator(get_discriminator_value),
+        ]
+
+    benchmark(rebuild_model, ThanksgivingDinner)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_tagged_union_with_callable_discriminator_nested_schema_generation(benchmark):
+    class BlackCat(BaseModel):
+        pet_type: Literal['cat']
+        color: Literal['black']
+        black_name: str
+
+    class WhiteCat(BaseModel):
+        pet_type: Literal['cat']
+        color: Literal['white']
+        white_name: str
+
+    Cat = Annotated[Union[BlackCat, WhiteCat], Field(discriminator='color')]
+
+    class Dog(BaseModel):
+        pet_type: Literal['dog']
+        name: str
+
+    class Model(DeferredModel):
+        pet: Annotated[Union[Cat, Dog], Field(discriminator='pet_type')]
+        n: int
+
+    benchmark(rebuild_model, Model)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_untagged_union_left_to_right_mode_schema_generation(benchmark):
+    class User(DeferredModel):
+        id: Union[str, int] = Field(union_mode='left_to_right')
+
+    benchmark(rebuild_model, User)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_untagged_union_smart_mode_schema_generation(benchmark):
+    class User(DeferredModel):
+        id: Union[int, str, UUID]
+        name: str
+
+    benchmark(rebuild_model, User)
+
+
+@pytest.mark.parametrize('field_type', StdLibTypes)
+@pytest.mark.benchmark(group='stdlib_schema_generation')
+def test_stdlib_type_schema_generation(benchmark, field_type):
+    class StdlibTypeModel(DeferredModel):
+        field: field_type
+
+    benchmark(rebuild_model, StdlibTypeModel)
```

Comment:
For this one, I'm not sure if we want to keep it as is :thinking: `StdLibTypes`  of ~60 types, which means 60 separate tests. This is going to clutter the codspeed output :face_with_diagonal_mouth:

Reply:
Ah true. I wonder if there's a way to group tests via codspeed so that it doesn't clutter them - would love to have some tests for std types so that we can gauge changes in performance for lots of different schema builds.

Reply:
Hi! I agree @Viicos . I researched Codspeed's API and found no trivial way of grouping the outputs. 

A solution might be to create categories/clusters of stlib types that are "similar" by some criteria and reduce the number of total benchmarks from 60+ to 3 to 5 benchmarks, each encapsulating a particular group.

An example of grouping would be:
```python
# ... existing code ...

StdLibTypes = [
    # Group 1: Basic Types
    [str, int, float, complex, bool, bytes],

    # Group 2: Date and Time
    [date, datetime, time, timedelta],

    # Group 3: Collections and Containers
    [deque, Deque[str], Deque[int], Deque[float], Deque[bytes],
     list, List[int], List[str], List[bytes], List[float],
     dict, Dict[str, float], Dict[str, bytes], Dict[str, int], Dict[str, str],
     tuple, Tuple[int, str, float],
     set, Set[int], Set[str],
     frozenset, FrozenSet[int], FrozenSet[str]],

    # Group 4: Special Types and Enums
    [Decimal, Color, ToolEnum, IPv4Address, IPv6Address, IPv4Interface, IPv6Interface,
     IPv4Network, IPv6Network, Path, Pattern, UUID, uuid4, uuid5, Point, User],

    # Group 5: Advanced Typing
    [Optional[int], Optional[str], Optional[float], Optional[bytes], Optional[bool],
     Sequence[int], Sequence[str], Sequence[bytes], Sequence[float],
     Iterable[int], Iterable[str], Iterable[bytes], Iterable[float],
     Callable[[int], int], Callable[[str], str],
     Literal['apple', 'pumpkin'], Type[Foo], Any]
]

# ... 
```

I also agree with @sydney-runkle about having a more granular view of the performance, depending on which types are involved. Maybe someone could point to a better way to organize the groups to account for this need.

Reply:
@Viicos and I had a chat - let's skip this benchmark for now (leave it in the code, just use a decorator to skip) - I think the granularity here is important - we can enable on branches that it makes sense to test on!

Reply:
Nice! Makes sense!